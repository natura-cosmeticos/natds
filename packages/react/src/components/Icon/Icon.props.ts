import { IconName } from '@naturacosmeticos/natds-icons'
import { Color, Size } from '@naturacosmeticos/natds-themes'

type IconSize = keyof Size
type IconColor = keyof Color
export interface IconProps {
  /**
   * Removes the entire element from the accessibility tree
   * when the Icon is purely decorative
   * @default true
   */
  ariaHidden?: boolean

  /**
  * The Icon color
  * @default "primary"
  */
  color?: IconColor;

  /**
  * The Icon name
  * @default "filled-default-mockup"
  */
  name?: IconName;

  /**
   * The Icon size
   * @default "standard"
   */
  size?: IconSize;

  /**
   * Optional ID for testing
   */
  testID?: string,
}
