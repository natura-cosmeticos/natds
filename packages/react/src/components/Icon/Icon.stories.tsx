import React from 'react'
import { Story, Meta } from '@storybook/react'
import { Icon, IconProps } from '.'

const componentStatus = `
---

**NOTE**: This component is available with the following attribute status:

- ✅ **Color**
    - Support Color tokens from theme
- ✅  **Size**
    - Support Size tokens from theme
- ✅  **Name**
    - Check all Icon names in [Icon Library](https://ds.natura.design/28db352be/p/94367e-icon-library/b/6154b9)

---
`

export default {
  title: 'Components/Icon',
  component: Icon,
  parameters: {
    componentSubtitle: 'A helper component to display icons from @naturacosmeticos/natds-icons package',
    docs: { description: { component: componentStatus } }
  }
} as Meta

export const Playground: Story<IconProps> = (args) => <Icon {...args} />
