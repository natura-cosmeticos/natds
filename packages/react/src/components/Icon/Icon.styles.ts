import { createUseStyles } from 'react-jss'
import { Theme } from '@naturacosmeticos/natds-themes'
import { IconProps } from './Icon.props'

type IconStyleProps = Required<Pick<IconProps, 'size' | 'color' >>

const styles = createUseStyles((theme: Theme) => ({
  icon: {
    color: ({ color }: IconStyleProps) => theme.color[color],
    fontSize: ({ size }: IconStyleProps) => theme.size[size],
    fontFamily: 'natds-icons',
    fontStyle: 'normal',
    userSelect: 'none'
  }
}))

export default styles
