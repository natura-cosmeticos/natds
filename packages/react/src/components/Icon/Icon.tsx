import React from 'react'
import iconNames from '@naturacosmeticos/natds-icons/dist/natds-icons.json'
import { IconProps } from './Icon.props'
import styles from './Icon.styles'

const Icon = ({
  ariaHidden = true,
  color = 'highlight',
  name = 'outlined-default-mockup',
  size = 'standard',
  testID = `icon-${name}`
}: IconProps): JSX.Element => {
  const unicodeName = iconNames[name].replace('%', '\\')
  const icon = JSON.parse(`["${unicodeName}"]`)[0]
  const classes = styles({ size, color })

  return (
    <i className={classes.icon} data-testid={testID} aria-hidden={ariaHidden}>
      {icon}
    </i>
  )
}

export default Icon
