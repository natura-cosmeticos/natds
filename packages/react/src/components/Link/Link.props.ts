import { IconProps } from '../Icon'

export type LinkVariants = 'regular' | 'underline'
export type LinkPositions = 'left' | 'right'
type TextSizes = 'heading1' | 'heading2' | 'heading3' | 'heading4' | 'heading5' | 'heading6' | 'subtitle1' | 'subtitle2' | 'body1' | 'body2' | 'caption' | 'overline'

export interface LinkProps {
  /**
   * Specify an optional className to be added to your Link
   */
  className?: string

  /**
   * Optional ID for testing
   */
  testID?: string;

  /**
   * The variant to use.
   */
  variant?: LinkVariants;

  /**
   * Specifies the URL of the page the link goes to.
   */
  href: React.AnchorHTMLAttributes<HTMLAnchorElement>['href'];

  /**
   * Specifies the language of the linked document.
   */
  hrefLang?: React.AnchorHTMLAttributes<HTMLAnchorElement>['hrefLang'];

  /**
   * The text to display inside the link.
   */
  text: string;

  /**
   * The size of the text.
   */
  textSize?: TextSizes;

  /**
   * The target for the link.
   */
  target?: React.AnchorHTMLAttributes<HTMLAnchorElement>['target'];

  /**
   * The position of the icon.
   */
  iconPosition?: LinkPositions;

  /**
   * The icon to display together with the text.
   */
  IconComponent?: React.ReactElement<IconProps>;

}
