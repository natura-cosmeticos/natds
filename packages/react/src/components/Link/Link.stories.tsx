import React from 'react'
import { Story, Meta } from '@storybook/react'
import { Link, LinkProps } from '.'
import { Icon } from '../Icon'

const componentStatus = `
---

**NOTE FOR UX**: This component is available in the following variants:

  - ✅ \`regular\`
  - ✅ \`underline\`

With the following attribute status:

  - **Icon**
      - ✅ \`left\`
      - ✅ \`right\`
  - **Color**
      - ✅ \`default\`
      - ✅ \`light\`

---
`

export default {
  title: 'Components/Link',
  component: Link,
  parameters: {
    componentSubtitle: 'The link component allows you to have a pattern for anchor text.',
    docs: { description: { component: componentStatus } }
  }
} as Meta

export const Playground: Story<LinkProps> = (args) => <Link {...args} />
Playground.args = {
  text: 'Link',
  href: 'https://'
}

export const Variants: Story<LinkProps> = (args) => (
  <>
    <Link {...args} variant="underline" text="underline variant example" />
    <Link {...args} variant="regular" text="regular variant example" />
  </>
)
Variants.args = { ...Playground.args }

export const Colors: Story<LinkProps> = (args) => (
  <>
    <Link {...args} text="default link color" />
    <Link {...args} textColor="light" text="light link color" />
  </>
)
Colors.args = { ...Playground.args }

export const WithIcon: Story<LinkProps> = (args) => (
  <>
    <Link
      {...args}
      text="right icon example"
      IconComponent={<Icon name="outlined-navigation-directionright-alt" color="link" />}
    />
    <Link
      {...args}
      text="left icon example"
      iconPosition="left"
      IconComponent={<Icon name="outlined-navigation-directionleft-alt" color="link" />}
    />
  </>
)
WithIcon.args = { ...Playground.args }
