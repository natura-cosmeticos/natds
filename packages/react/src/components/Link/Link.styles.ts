import { createUseStyles } from 'react-jss'
import { Theme } from '@naturacosmeticos/natds-themes'
import { LinkProps } from './Link.props'

type LinkStyleProps = Required<Pick<LinkProps, 'variant'>>

const styles = createUseStyles((theme: Theme) => ({
  link: {
    color: theme.color.link,
    fontFamily: theme.typography.fontFamily.primary,
    fontSize: theme.body1.fontSize,
    letterSpacing: theme.body1.letterSpacing,
    lineHeight: theme.body1.lineHeight,
    textDecoration: ({ variant }: LinkStyleProps) => (variant === 'underline' ? 'underline' : 'none')
  }
}))

export default styles
