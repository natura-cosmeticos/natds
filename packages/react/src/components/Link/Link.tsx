import React from 'react'
import { LinkProps } from './Link.props'
import styles from './Link.styles'

const Link = React.forwardRef<HTMLAnchorElement, LinkProps>(
  ({
    className = '',
    testID,
    variant = 'regular',
    href,
    hrefLang,
    text,
    target,
    ...props
  }, ref) => {
    const { link } = styles({ variant })

    return (
      <a
        className={`${className} ${link}`}
        data-testid={testID}
        ref={ref}
        href={href}
        hrefLang={hrefLang}
        target={target}
        {...props}
      >
        {text}
      </a>
    )
  }
)

export default Link
