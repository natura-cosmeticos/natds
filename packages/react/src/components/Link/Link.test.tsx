import * as React from 'react'
import Link from '.'
import renderWithTheme from '../../helpers/renderWithTheme'
import { Icon } from '../Icon'

const defaultProps = {
  text: 'Link',
  href: 'https://'
}

describe('Link component', () => {
  it('should render correctly with default props', () => {
    const { styles, component } = renderWithTheme(<Link {...defaultProps} />)

    expect([styles.toString(), component.container]).toMatchSnapshot()
  })
  it('should render correctly when the variant is regular', () => {
    const { styles, component } = renderWithTheme(<Link {...defaultProps} variant="regular" />)

    expect([styles.toString(), component.container]).toMatchSnapshot()
  })

  it('should render correctly when the textColor is light', () => {
    const { styles, component } = renderWithTheme(<Link {...defaultProps} textColor="light" />)

    expect([styles.toString(), component.container]).toMatchSnapshot()
  })

  it('should render correctly when IconComponent props with iconPosition left', () => {
    const { styles, component } = renderWithTheme(
      <Link
        {...defaultProps}
        iconPosition="left"
        IconComponent={
          <Icon name="filled-default-mockup" />
        }
      />
    )

    expect([styles.toString(), component.container]).toMatchSnapshot()
  })
})
