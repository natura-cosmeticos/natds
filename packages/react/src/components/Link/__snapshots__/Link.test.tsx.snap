// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link component should render correctly when the textColor is light 1`] = `
Array [
  ".container-0-2-1 {
  width: fit-content;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: center;
}
.link-0-2-2 {
  font-size: 16px;
  font-family: Roboto, Arial;
  line-height: 1.5;
  letter-spacing: 0.51px;
}
.link-0-2-2:hover {}
.container-d0-0-2-3 {
  flex-direction: row;
}
.link-d1-0-2-4 {
  color: #bbbbbb;
  text-decoration: underline;
  margin-right: 8px;
}
.link-d1-0-2-4:hover {
  text-decoration: none;
}",
  <div>
    <span
      class="container-0-2-1 container-d0-0-2-3"
    >
      <a
        class=" link-0-2-2 link-d1-0-2-4"
        href="https://"
      >
        Link
      </a>
    </span>
  </div>,
]
`;

exports[`Link component should render correctly when the variant is regular 1`] = `
Array [
  ".container-0-2-1 {
  width: fit-content;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: center;
}
.link-0-2-2 {
  font-size: 16px;
  font-family: Roboto, Arial;
  line-height: 1.5;
  letter-spacing: 0.51px;
}
.link-0-2-2:hover {}
.container-d0-0-2-3 {
  flex-direction: row;
}
.link-d1-0-2-4 {
  color: #227bbd;
  text-decoration: none;
  margin-right: 8px;
}
.link-d1-0-2-4:hover {
  text-decoration: underline;
}",
  <div>
    <span
      class="container-0-2-1 container-d0-0-2-3"
    >
      <a
        class=" link-0-2-2 link-d1-0-2-4"
        href="https://"
      >
        Link
      </a>
    </span>
  </div>,
]
`;

exports[`Link component should render correctly with IconComponent rendered to the left 1`] = `
Array [
  ".container-0-2-1 {
  width: fit-content;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: center;
}
.link-0-2-2 {
  font-size: 16px;
  font-family: Roboto, Arial;
  line-height: 1.5;
  letter-spacing: 0.51px;
}
.link-0-2-2:hover {}
.container-d0-0-2-3 {
  flex-direction: row-reverse;
}
.link-d1-0-2-4 {
  color: #227bbd;
  text-decoration: underline;
  margin-left: 8px;
}
.link-d1-0-2-4:hover {
  text-decoration: none;
}
.icon-0-2-5 {
  font-family: natds-icons;
  user-select: none;
  pointer-events: none;
}
.icon-d0-0-2-6 {
  color: #000000;
  font-size: 24px;
}",
  <div>
    <span
      class="container-0-2-1 container-d0-0-2-3"
    >
      <a
        class=" link-0-2-2 link-d1-0-2-4"
        href="https://"
      >
        Link
      </a>
      <i
        aria-hidden="true"
        class=" icon-0-2-5 icon-d0-0-2-6 natds-icons natds-icons-filled-default-mockup"
        data-testid="icon-filled-default-mockup"
        role="img"
      />
    </span>
  </div>,
]
`;

exports[`Link component should render correctly with IconComponent rendered to the right by default 1`] = `
Array [
  ".container-0-2-1 {
  width: fit-content;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: center;
}
.link-0-2-2 {
  font-size: 16px;
  font-family: Roboto, Arial;
  line-height: 1.5;
  letter-spacing: 0.51px;
}
.link-0-2-2:hover {}
.container-d0-0-2-3 {
  flex-direction: row;
}
.link-d1-0-2-4 {
  color: #227bbd;
  text-decoration: underline;
  margin-right: 8px;
}
.link-d1-0-2-4:hover {
  text-decoration: none;
}
.icon-0-2-5 {
  font-family: natds-icons;
  user-select: none;
  pointer-events: none;
}
.icon-d0-0-2-6 {
  color: #000000;
  font-size: 24px;
}",
  <div>
    <span
      class="container-0-2-1 container-d0-0-2-3"
    >
      <a
        class=" link-0-2-2 link-d1-0-2-4"
        href="https://"
      >
        Link
      </a>
      <i
        aria-hidden="true"
        class=" icon-0-2-5 icon-d0-0-2-6 natds-icons natds-icons-filled-default-mockup"
        data-testid="icon-filled-default-mockup"
        role="img"
      />
    </span>
  </div>,
]
`;

exports[`Link component should render correctly with default props 1`] = `
Array [
  ".container-0-2-1 {
  width: fit-content;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: center;
}
.link-0-2-2 {
  font-size: 16px;
  font-family: Roboto, Arial;
  line-height: 1.5;
  letter-spacing: 0.51px;
}
.link-0-2-2:hover {}
.container-d0-0-2-3 {
  flex-direction: row;
}
.link-d1-0-2-4 {
  color: #227bbd;
  text-decoration: underline;
  margin-right: 8px;
}
.link-d1-0-2-4:hover {
  text-decoration: none;
}",
  <div>
    <span
      class="container-0-2-1 container-d0-0-2-3"
    >
      <a
        class=" link-0-2-2 link-d1-0-2-4"
        href="https://"
      >
        Link
      </a>
    </span>
  </div>,
]
`;
